### PR TITLE
Connector Abstraction

### LINKED ISSUE #1898: Abstract single module definition of connector
In a single trait, we need to include all the types and methods required to _completely specify_ a connector. Then we need to implement the trait for all existing connectors, and modify a connector integration so that it is as simple as adding a new connector implementation to a list of connectors.

The goal is to make it _crystal clear_ for a connector author _exactly_ what is necessary to implement in order to have a fully-working connector, integrated with the regression tests, standardized tests, etc.

### PR DESCRIPTION
**Pull request by djspiewak**
Adds the `BackendModule` abstraction without actually wiring it up to anything.  A stub marklogic backend is currently provided.  Work to fully wire up marklogic is preceding apace in a derived branch, but this PR is necessary to avoid perpetuating merge conflicts with @rabbitonweb.

Fixes #1898

### CODE REVIEW

File: connector/src/main/scala/quasar/connector/BackendModule.scala [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  private final implicit def _ExtractPathQSM[T[_[_]]: RecursiveT]: ExtractPath[QSM[T, ?], APath] = ExtractPathQSM
+  private final implicit def _QSCoreInject[T[_[_]]] = QSCoreInject[T]
+  private final implicit def _MonadM = MonadM
+  private final implicit def _MonadFsErrM = MonadFsErrM
+  private final implicit def _PhaseResultTellM = PhaseResultTellM
+  private final implicit def _PhaseResultListenM = PhaseResultListenM
+  private final implicit def _UnirewriteT[T[_[_]]: BirecursiveT: OrderT: EqualT: ShowT: RenderTreeT] = UnirewriteT[T]
+  private final implicit def _UnicoalesceCap[T[_[_]]: BirecursiveT: OrderT: EqualT: ShowT: RenderTreeT] = UnicoalesceCap[T]
+
+  final def definition[Task] = FileSystemDef fromPF {
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/connector/BackendModule.scala]: Doesn't look like the `Task` type variable is necessary.

Comment (djspiewak) [inline connector/src/main/scala/quasar/connector/BackendModule.scala]: You're quite correct.  All it does is create the potential for an incredibly confusing type error down the line.

**Comment by djspiewak** [comment]
@wemrysi I made some changes to address your comments.  Let me know if things look good.

Earth to @sellout?

@rabbitonweb Can we land this in master today?

**Comment by djspiewak** [comment]
Since @rabbitonweb is not responding, I'm reassigning this to @wemrysi.  This needs to be merged for work that is beginning today on FBE.  (well, I guess *need* is a strong word; I can work around it with Git shenanigans, but no one likes Git shenanigans)

**Comment by alissapajer** [comment]
@sellout I think this is now pending on review from you. Could you take a look?

**Pull request merged**